File: src/main/java/com/hazza/LeetCodeByJava/No11ContainerWithMostWater.java
package com.hazza.LeetCodeByJava;

/**
 * https://leetcode.com/problems/container-with-most-water/#/description
 */
public class No11ContainerWithMostWater {
    public int maxArea(int[] height) {
        int res = 0, left = 0, right = height.length - 1;
        while (left != right) {
            res = Math.max(res, (right - left) * Math.min(height[left], height[right]));
            if (height[left] < height[right]) ++left;
            else --right;
        }
        return res;
    }

}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No11_Container_With_Most_Water.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/9 20:33
# https://leetcode.com/problems/container-with-most-water/
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        res, left, right = 0, 0, len(height) - 1

        while left != right:
            res = max(res, (right - left) * min(height[left], height[right]))
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return res


if __name__ == '__main__':
    solution = Solution()
    assert solution.maxArea([1,8,6,2,5,4,8,3,7]) == 49